### PR TITLE
Blocks: Create an Audio block on audio file drop

### LINKED ISSUE #8020: Blocks: Dropping an audio file should insert an Audio block
Related: #8021

When drag-dropping an audio file from my computer, a File block is created. Since Audio block exists, it should include a transform to create itself from the file (via the `transforms` property `files` type).

See also:

- [Image block reference implementation](https://github.com/WordPress/gutenberg/blob/b06a33569a2b860110cb8d3f113ccf6c2b36e492/core-blocks/image/index.js#L128-L144)
- [Block API `transforms` property](https://wordpress.org/gutenberg/handbook/block-api/#transforms-optional)

### PR DESCRIPTION
**Pull request by Copons**
## Description

Fix #8020

Automatically create an Audio block when drag-and-dropping an audio file from the computer.

## How has this been tested?

Environment: local Docker
Tested manually and via `npm test`.

## Screenshots 

jul-19-2018 14-48-57

If the upload fails, the block will display the `MediaPlaceholder` with an error notice containing the upload error.
I'm not entirely sure how this will be localized though.
Also: the idea is to immediately empty the `src` attribute of the audio file, in order to remove the `blob:` filename, but for some reasons it takes quite a while for `setAttribute` to update the placeholder.

screen shot 2018-07-19 at 16 51 17

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: core-blocks/audio/edit.js [hunk 1]
```diff
@@ -36,6 +38,30 @@ class AudioEdit extends Component {
...
+			if ( file ) {
+				editorMediaUpload( {
+					filesList: [ file ],
+					onFileChange: ( [ { url } ] ) => {
+						setAttributes( { src: url } );
+					},
+					onError: ( e ) => {
+						setAttributes( { src: undefined, id: undefined } );
+						this.setState( { editing: true } );
+						noticeOperations.createErrorNotice( e );
```
Comment (jorgefilipecosta) [inline core-blocks/audio/edit.js]: I like the mechanism used to display error notices which are not displayed in other cases. We should probably update existing code to also offer this possibility.

File: core-blocks/audio/edit.js [hunk 2]
```diff
@@ -36,6 +38,30 @@ class AudioEdit extends Component {
 		this.onSelectURL = this.onSelectURL.bind( this );
 	}
 
+	componentDidMount() {
+		const { attributes, noticeOperations, setAttributes } = this.props;
+		const { id, src = '' } = attributes;
+
+		if ( ! id && src.indexOf( 'blob:' ) === 0 ) {
```
Comment (jorgefilipecosta) [inline core-blocks/audio/edit.js]: Unfortunately using this logic to decide when to upload is vulnerable to our undo problem.
If after the upload we undo, as the set of the attributes to the blob and to the final URL are two different levels we create two undo levels.
When the user undoes the upload, we will get back to the state where we have a blob URL so the upload will happen again making the undo operation useless. It is possible to see the problem by uploading an audio block, waiting for the upload to happen, pressing undo, waiting a little bit, pressing undo again, waiting, pressing undo again and checking we are blocked.

This a complex problem related to undo for which we did not decide yet the way to attack it https://github.com/WordPress/gutenberg/pull/6325#issuecomment-402242054.

I think that maybe for this cases, (the gallery has similar code in the shortcode to gallery transform) we should use the state of the component to flag that the upload/request was made and in that case, the component does not repeat the action.

Comment (jorgefilipecosta) [reply core-blocks/audio/edit.js]: I also think storing the blobUrl as an attribute is ok. What I was suggesting is just having a flag ( e.g: uploadCompleted ) in the state so we don't repeat the upload.

Comment (Copons) [reply core-blocks/audio/edit.js]: > It is possible to see the problem by uploading an audio block, waiting for the upload to happen, pressing undo, waiting a little bit, pressing undo again, waiting, pressing undo again and checking we are blocked.

@jorgefilipecosta 🤔 I'm seeing a slightly different behaviour.

I drop an audio file, the `AudioEdit` gets a blob src, and eventually obtains a proper file src.

Now, I hit Undo.
The `AudioEdit` loses the proper file src and gets back the blob one.
And that's it. No new upload starts now.

Hitting Undo again, the Audio component loses the blob src too, and it's removed altogether.
If I hit Redo now, the Audio component will get the blob src back, and it will trigger a new upload (the new audio file has an incremented ID).

I guess this happens because the autoupload only happens on `componentDidMount`, which is not triggered on the first Undo (while it's indeed triggered on Undo-Undo-Redo, because the component is unmounted on the second Undo).

In other words, I don't think an `uploadCompleted` would really help us:
- On first Undo, the upload does not start again anyway.
- On Undo-Undo-Redo, the component is unmounted and mounted again from scratch, losing the internal state, so the file would be re-uploaded anyway.

Comment (jorgefilipecosta) [reply core-blocks/audio/edit.js]: Hi @Copons, sorry for the delay in answering this notification got lost in the middle of notifications received while AFK.
I tried to retest this but I'm having problems executing this branch 
```
Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings. 
```
I think its related to other changes made in parallel, would it be possible to rebase this branch so we can have a new look?
Thank you in advance.

**Comment by Copons** [comment]
@jorgefilipecosta rebased and fixed a function name change (`editorMediaUpload` has been renamed into `mediaUpload`). Thanks for 👀 this!

**Review by jorgefilipecosta — APPROVED** [review]
I did new set of tests and things looked great. I was able to use undo without getting locked. To undo the audio upload we need to press undo two times (one for the file it self other for the block) but that is expected in our current implementation and a solution is being worked on.
I think we can merge this, thank you for your contribution @Copons!
